src/test/ui/rfc-2008-non-exhaustive/variant.rs

```diff
@@ -0,0 +1,49 @@
+// aux-build:variants.rs
+
+extern crate variants;
+
+use variants::NonExhaustiveVariants;
+
+fn main() {
+    let variant_struct = NonExhaustiveVariants::Struct { field: 640 };
+    //~^ ERROR cannot create non-exhaustive variant
+
+    let variant_tuple = NonExhaustiveVariants::Tuple(640);
+    //~^ ERROR tuple variant `Tuple` is private [E0603]
+
+    let variant_unit = NonExhaustiveVariants::Unit;
+    //~^ ERROR unit variant `Unit` is private [E0603]
+
+    match variant_struct {
+        NonExhaustiveVariants::Unit => "",
+        NonExhaustiveVariants::Tuple(fe_tpl) => "",
+        //~^ ERROR `..` required with variant marked as non-exhaustive
+        NonExhaustiveVariants::Struct { field } => ""
+        //~^ ERROR `..` required with variant marked as non-exhaustive
+    };
+
+    if let NonExhaustiveVariants::Tuple(fe_tpl) = variant_struct {
+        //~^ ERROR `..` required with variant marked as non-exhaustive
+    }
+
+    if let NonExhaustiveVariants::Struct { field } = variant_struct {
+        //~^ ERROR `..` required with variant marked as non-exhaustive
+    }
+
+    // Everything below this is expected to compile successfully.
+
+    match variant_struct {
+        NonExhaustiveVariants::Unit => "",
+        NonExhaustiveVariants::Tuple(fe_tpl, ..) => "",
```

Comment:
These two variants are not supposed to work.
The point of lowering visibility to `pub(crate)` is that it makes constructors entirely unobservable from other crates in natural way.
You can, e.g. turn your non-exhaustive unit variant into a non-unit variant and users certainly won't break, the "unit-ness" becomes an internal convenience, an implementation details.

I guess something like `NonExhaustiveVariants::Tuple(prefix_fields, ..)` may be supported given enough demand (if we are sure that defining numeric fields explicitly will never be allowed), but at cost of some hacks (as you could notice with the resolve changes).

Anyway, I'd prefer not to do this now.
So, could you revert the changes in `librustc_resolve/lib.rs`, `librustc_resolve/resolve_imports.rs`, and `librustc_typeck/check/_match.rs` for tuple struct patterns.
(\+ some commit squashing after that would be nice.)

Reply:
There was some more detailed Zulip conversation with eddyb about this, but I can't find it right now.

Reply:
Here's the [aforementioned Zulip conversation](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/rfc-2008/near/132662043).

Reply:
Fixed this.